presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java

```diff
@@ -0,0 +1,468 @@
+                RowExpression newPredicate = remapExpression(predicate, varMap);
+                context.addCteConsumerInfo(cteName, producerColumns, ImmutableList.of(newPredicate));
+                return null;
+            }
+
+            return super.visitFilter(node, context);
+        }
+
+        public Void visitProject(ProjectNode node, CteContext context)
+        {
+            if (isCteConsumerFilterRestrict(node)) {
+                // get predicate and used columns
+                CteConsumerNode cteConsumerNode = extractCteConsumer(node);
+                Map<VariableReferenceExpression, VariableReferenceExpression> varMap = constructConsumerToProducerVarMap(cteConsumerNode, context);
+                List<VariableReferenceExpression> usedColumns =
+                        node.getAssignments().getExpressions().stream().map(expression -> (VariableReferenceExpression) remapExpression(expression, varMap)).collect(Collectors.toList());
+
+                FilterNode filterNode = extractFilterNode(node);
+                RowExpression predicate;
+                if (filterNode != null) {
+                    predicate = remapExpression(filterNode.getPredicate(), varMap);
+                    // extract predicate columns and add to used columns
+                    usedColumns.addAll(VariablesExtractor.extractAll(predicate));
+                }
+                else {
+                    predicate = constant(true, BOOLEAN);
+                }
+
+                context.addCteConsumerInfo(cteConsumerNode.getCteName(), usedColumns, ImmutableList.of(predicate));
+                return null;
+            }
+
+            return super.visitProject(node, context);
+        }
+
+        @Override
+        public Void visitCteConsumer(CteConsumerNode node, CteContext context)
+        {
+            // if we reach this point, it means that the CTE consumer had no filter or projection on top of it and we must take all columns and rows
+            String cteName = node.getCteName();
+            // TODO: support pushing of projections in the cte consumer (similar to table scan projection push down)
+            // for now, take the original columns of the CTE producer
+            List<VariableReferenceExpression> producerColumns = context.getCteProducerColumns(cteName);
+            checkState(producerColumns != null, "No producer with name " + cteName + " found");
+            // no filter encountered on top of this consumer: all rows must be read
+            RowExpression predicate = constant(true, BOOLEAN);
+            context.addCteConsumerInfo(cteName, producerColumns, ImmutableList.of(predicate));
+            return null;
+        }
+
+        @Override
+        public Void visitSequence(SequenceNode node, CteContext context)
+        {
+            List<PlanNode> cteProducers = node.getCteProducers();
+            for (int i = cteProducers.size() - 1; i >= 0; i--) {
```

Comment:
Do we need to iterate in reverse order? If not, we can use utils like `forEach`

Reply:
we do need it like that, ctes may define new ctes used in subsequent cte nodes, and at the moment we've chosen to store them in this order. We may in the future refactor the representation to store ctes from the inner-most to outer-most scope in the right order.
CC: @jaystarshot

Reply:
Yeah we need to refactor that logic for scheduling improvements too, currently the execution follows reverse topological order